design/20210209.certificates.k8s.io-adoption.md

```diff
@@ -0,0 +1,292 @@
+resource.
+
+### Read all access to CertificateSigningRequests
+
+`CertificateSigningRequests` are cluster scoped resources, whereas
+`CertificateRequests` are not. This means that in order for users to have the
+same debugging and usage capabilities of the `CertificateRequest`, users will
+have the ability to list all requests in the cluster, rather than being scoped to
+specific namespaces. This is a regression in security.
+
+A solution to this problem could be given to users by providing examples or
+guides to external tooling (e.g. [Open Policy Agent
+(OPA)](https://github.com/open-policy-agent/opa), whereby cluster administrators
+can limit users access to certain `CertificateSigningRequests` based on label
+selectors.
+
+
+### Namespaced issuers referenced by CertificateSigningRequest
+
+`CertificateSigningRequests` are cluster scoped resources, whereas
+`CertificateRequests` are not. cert-manager has the concept of of namespaced
+issuers that may only be referenced by `CertificateRequests` that reside in
+the same namespace. Since `CertificateSigningRequests` can reference any
+namespaced `Issuer`, there must be some mechanism to prevent privilege
+escalation (Alice references an issuer in Bob's namespace).
+
+cert-manager will enforce an
+[RBAC](https://kubernetes.io/docs/reference/access-authn-authz/rbac/) noun and
+verb whereby the requester must have this role bound to them, in order for the
+`CertificateSigningRequest` referencing a namespaced `Issuer` be approved by the
+cert-manager controller. See [here](#conditions).
+
+This will be done via a
+[`SubjectAccessReview`](https://github.com/kubernetes/api/blob/4a626d306b987a4096cf0784ec01af1be2f6d67f/authorization/v1/types.go#L52)
+against the `UserInfo` fields on the `CertificateSigningRequest`. The required
+role must be of `certificates.k8s.io/CertificateSigningRequest` with the verb
+`create` within the target namespace.
+
+Since `Certificates` are also namespaced resources, cert-manager validates that
+created `Certificate` resources may only reference `Issuers` in the same
+namespace, or `ClusterIssuers`. Any resulting `CertificateSigningRequests` will
+reference `Issuers` in the same namespace.
+
+### Duration Field
+
+`CertificateSigningRequests` do not include a `duration` field. To have parity
+with the `CertificateRequest` resource, the duration field will be moved to an
+annotation as `cert-manager.io/duration` which includes a [Go time duration
+string](https://golang.org/pkg/time/#Duration.String). This annotation may not
+be changed after creation.
+
+### CA Field
+
+`CertificateSigningRequests` do not include a `ca` field. To have parity with
+the `CertificateRequest` resource, the `ca` field will be moved to an annotation
+as `cert-manager.io/ca`. This annotation value contains the base 64 PEM encoded
+CA certificate (if one was returned by the signer). This annotation may not be
+changed after creation.
```

Comment:
Do we mean that we will have some webhook running for making sure the `cert-manager.io/ca` annotation is not mutated? Weren't we trying to avoid having to rely on webhooks due to the fact that webhooks are security-sensitive and are managed by operations teams as opposed to dev teams?

Reply:
The ca field is now no longer present.